python/lsst/daf/butler/cli/butler.py

```diff
@@ -0,0 +1,89 @@
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import click
+import logging
+import os
+import sys
+
+from lsst.daf.butler.cli.opt.verbose import verbose_option
+import lsst.log
+from lsst.utils import doImport
+
+
+def _getPluginList():
+    pluginModules = os.environ.get("DAF_BUTLER_PLUGINS")
+    if pluginModules is None:
+        return None
+    return pluginModules.split(":")
+
+
+def _importPlugins(pluginList, group):
+    for pluginName in pluginList:
+        try:
+            plugin = doImport(pluginName)
+        except (TypeError, ModuleNotFoundError, ImportError) as err:
+            logging.warning("Could not import plugin from %s, skipping.", pluginName)
+            logging.debug("Plugin import exception: %s", err)
+            continue
+        for command in plugin.__all__:
+            fullCommand = pluginName + "." + command
+            logging.debug("Loading CLI command: %s", fullCommand)
+            try:
+                cmd = doImport(fullCommand)
+            except (TypeError, ModuleNotFoundError, ImportError) as err:
+                logging.warning("Could not import command from %s, skipping.", fullCommand)
+                logging.debug("Command import exception: %s", err)
+                continue
+            try:
+                group.add_command(cmd)
+            except Exception as err:
+                logging.warning("Failed while adding command %s from %s, skipping.", command, pluginName)
+                logging.debug("Exception during add_command: %s", err)
+
+
+def _addCommands(group):
+    pluginList = _getPluginList()
+    _importPlugins(pluginList, group)
+
+
+@click.group()
+# this group doesn't use the verbose command but does allow (ingore) it,
+# because the -v flag is used for logging while building the group.
+@verbose_option
+def cli(verbose):
+    pass
+
+
+lgr = logging.getLogger("lsst.daf.butler")
+log = lsst.log.Log.getLogger("lsst.daf.butler")
+
+
+def getCli():
+    _addCommands(cli)
+    return cli
+
+
+def main():
+    verbose = len(sys.argv) > 1 and "-v" == sys.argv[1]
+    logging.basicConfig(level=logging.DEBUG if verbose else logging.INFO)
+    log.setLevel(lsst.log.Log.DEBUG if verbose else lsst.log.Log.INFO)
+    return getCli()()
```

Comment:
This file doesn't *have* to be called butler.py; it's called by the file at `bin/butler`, and could be named e.g. `butlerCli.py` if that could reduce confusion between this and the file that owns the Butler class (`_butler.py`). I wasn't sure which way to go on that.